File: bot-api/python/src/robocode_tank_royale/bot_api/internal/bot_handshake_factory.py
import os
import sys
from ..bot_info import BotInfo
from ..mapper import InitialPositionMapper
from .env_vars import EnvVars

from robocode_tank_royale.schema import BotHandshake
from robocode_tank_royale.schema.message import Message


class BotHandshakeFactory:
    @staticmethod
    def create(
        session_id: str, bot_info: BotInfo, is_droid: bool, secret: "str | None"
    ) -> BotHandshake:
        handshake = BotHandshake(
            session_id=session_id,
            type=Message.Type.BOT_HANDSHAKE,  # type: ignore
            name=bot_info.name,
            version=bot_info.version,
            authors=list(bot_info.authors),
            description=bot_info.description,  # type: ignore
            homepage=bot_info.homepage,  # type: ignore
            country_codes=list(bot_info.country_codes),
            game_types=list(bot_info.game_types),
            platform=bot_info.platform,
            programming_lang=bot_info.programming_lang,  # type: ignore
            initial_position=InitialPositionMapper.map(bot_info.initial_position),  # type: ignore
            team_id=EnvVars.get_team_id(),  # type: ignore
            team_name=EnvVars.get_team_name(),  # type: ignore
            team_version=EnvVars.get_team_version(),  # type: ignore
            is_droid=is_droid,
            secret=secret,
        )

        # Set debugger_attached field (ADR-0035)
        debugger_attached = BotHandshakeFactory.is_debugger_attached()
        handshake.debugger_attached = debugger_attached

        # Log hint if debugger is detected
        if debugger_attached:
            print(
                "Debugger detected. Consider enabling breakpoint mode for this bot in the controller."
            )

        return handshake

    @staticmethod
    def is_debugger_attached() -> bool:
        """
        Detects if a debugger is attached to the process.
        @return true if a debugger is attached, false otherwise
        """
        # Check for ROBOCODE_DEBUG environment variable override
        env = os.environ.get("ROBOCODE_DEBUG", "").lower()
        if env == "true":
            return True
        if env == "false":
            return False

        # Check for trace function (pydevd, debugpy, etc.)
        if sys.gettrace() is not None:
            return True

        # Check for specific debugger modules
        return "debugpy" in sys.modules or "pydevd" in sys.modules

File: bot-api/dotnet/api/src/internal/BotHandshakeFactory.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Robocode.TankRoyale.BotApi.Mapper;
using Robocode.TankRoyale.BotApi.Util;
using Robocode.TankRoyale.Schema;

namespace Robocode.TankRoyale.BotApi.Internal;

static class BotHandshakeFactory
{
    internal static BotHandshake Create(string sessionId, BotInfo botInfo, bool isDroid, string serverSecret)
    {
        var handshake = new BotHandshake
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.BotHandshake),
            SessionId = sessionId,
            Name = botInfo.Name,
            Version = botInfo.Version,
            Authors = new List<string>(botInfo.Authors),
            Description = botInfo.Description,
            Homepage = botInfo.Homepage,
            CountryCodes = botInfo.CountryCodes != null
                ? new List<string>(botInfo.CountryCodes)
                : new List<string>(),
            GameTypes = botInfo.GameTypes != null ? new HashSet<string>(botInfo.GameTypes) : new HashSet<string>(),
            Platform = botInfo.Platform,
            ProgrammingLang = botInfo.ProgrammingLang,
            InitialPosition = InitialPositionMapper.Map(botInfo.InitialPosition),
            TeamId = EnvVars.GetTeamId(),
            TeamName = EnvVars.GetTeamName(),
            TeamVersion = EnvVars.GetTeamVersion(),
            IsDroid = isDroid,
            Secret = serverSecret,
        };

        // Set DebuggerAttached field
        bool debuggerAttached = IsDebuggerAttached();
        handshake.DebuggerAttached = debuggerAttached;

        // Log hint if debugger is detected
        if (debuggerAttached)
        {
            System.Console.WriteLine("Debugger detected. Consider enabling breakpoint mode for this bot in the controller.");
        }

        return handshake;
    }

    /// <summary>
    /// Detects if a debugger is attached to the process.
    /// </summary>
    /// <returns>true if a debugger is attached, false otherwise</returns>
    private static bool IsDebuggerAttached()
    {
        // Check for ROBOCODE_DEBUG environment variable override
        var env = Environment.GetEnvironmentVariable("ROBOCODE_DEBUG");
        if ("true".Equals(env, StringComparison.OrdinalIgnoreCase))
            return true;
        if ("false".Equals(env, StringComparison.OrdinalIgnoreCase))
            return false;

        // Check if a managed debugger is attached
        return Debugger.IsAttached;
    }
}